astropy/table/table.py

```diff
@@ -742,20 +742,35 @@ def _set_masked_from_cols(self, cols):
                 self._set_masked(True)
 
     def _init_from_list_of_dicts(self, data, names, dtype, n_cols, copy):
-        names_from_data = set()
-        for row in data:
-            names_from_data.update(row)
-
-        cols = {}
-        for name in names_from_data:
-            cols[name] = []
-            for i, row in enumerate(data):
-                try:
-                    cols[name].append(row[name])
-                except KeyError:
-                    raise ValueError('Row {0} has no value for column {1}'.format(i, name))
-        if all(name is None for name in names):
-            names = sorted(names_from_data)
+        if isinstance(data[0], OrderedDict):
+            names_from_data = list(data[0].keys())
+
+            cols = {}
```

Comment:
I would remove the repetitive part from the conditional and rather have another short conditional that changes `names_from_data` to a shorted list in case it's a set.

Reply:
OK, I changed this!